src/subhalo_tracking.cpp

```diff
@@ -853,28 +869,46 @@ class SubhaloMasker_t
   {
     ExclusionList.reserve(np_guess);
   }
-  void Mask(HBTInt subid, vector<Subhalo_t> &Subhalos)
+  void Mask(HBTInt subid, vector<Subhalo_t> &Subhalos, int SnapshotIndex)
   {
     auto &subhalo = Subhalos[subid];
     for (auto nestedid :
          subhalo
            .NestedSubhalos) // TODO: do we have to do it recursively? satellites are already masked among themselves?
-      Mask(nestedid, Subhalos);
+      Mask(nestedid, Subhalos, SnapshotIndex);
 
     if (subhalo.Nbound <= 1)
       return; // skip orphans
 
+    if (subhalo.SnapshotIndexOfBirth == SnapshotIndex)
+      return; // skip newly created centrals
+
+    // If all tracers are removed during masking, we need to add one back
+    bool hasTracer = false;
+    // Save the most bound tracer from the previous snapshot so we can add it back
+    auto tracer = subhalo.Particles[subhalo.GetTracerIndex()];
     auto it_begin = subhalo.Particles.begin(), it_save = it_begin;
     for (auto it = it_begin; it != subhalo.Particles.end(); ++it)
     {
       auto insert_status = ExclusionList.insert(it->Id);
       if (insert_status.second) // inserted, meaning not excluded
       {
+        if (!hasTracer && it->IsTracer())
```

Comment:
Do we need two conditions here?

Reply:
Practically it won't make a difference, but I think this is slightly more optimised